presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcPageSource.java

```diff
@@ -164,27 +159,50 @@ public Page getNextPage()
 
             long filePosition = recordReader.getFilePosition();
 
+            // lazy load rowsDeleted until getNextPage call
+            // rowsDeleted will only be loaded once
+            // for every page, will generate its rowsToKeep
+            Optional<BitSet> rowsDeleted = deltaShardLoader.get();
+            RowsToKeepResult rowsToKeep = loadRowsToKeep(rowsDeleted, batchSize, filePosition);
+
             Block[] blocks = new Block[columnIndexes.length];
             for (int fieldId = 0; fieldId < blocks.length; fieldId++) {
                 if (constantBlocks[fieldId] != null) {
-                    blocks[fieldId] = constantBlocks[fieldId].getRegion(0, batchSize);
+                    blocks[fieldId] = constantBlocks[fieldId].getRegion(0, rowsToKeep.isKeepAll ? batchSize : rowsToKeep.size());
                 }
                 else if (columnIndexes[fieldId] == ROWID_COLUMN) {
-                    blocks[fieldId] = buildSequenceBlock(filePosition, batchSize);
+                    blocks[fieldId] = buildSequenceBlock(filePosition, batchSize, rowsToKeep);
                 }
                 else {
-                    blocks[fieldId] = new LazyBlock(batchSize, new OrcBlockLoader(columnIndexes[fieldId]));
+                    blocks[fieldId] = new LazyBlock(batchSize, new OrcBlockLoader(columnIndexes[fieldId], rowsToKeep));
                 }
             }
 
-            return new Page(batchSize, blocks);
+            return new Page(rowsToKeep.isKeepAll ? batchSize : rowsToKeep.size(), blocks);
         }
         catch (IOException | RuntimeException e) {
             closeWithSuppression(e);
             throw new PrestoException(RAPTOR_ERROR, e);
         }
     }
 
+    private RowsToKeepResult loadRowsToKeep(Optional<BitSet> rowsDeleted, int batchSize, long filePosition)
+    {
+        if (rowsDeleted.isPresent() && rowsDeleted.get().cardinality() > 0) {
```

Comment:
Do we have unit test to cover the if body? I didn't see any somehow

Reply:
Yes,  ``TestOrcStorageManager`` will test the functionality of reading the rowsToKeep result.
like ``testWriteDeltaDelete``, ``testWriteDeltaDeleteMerge``.